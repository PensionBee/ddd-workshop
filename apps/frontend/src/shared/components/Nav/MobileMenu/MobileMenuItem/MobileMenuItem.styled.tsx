import { theme } from "@shared/styling/theme";
import styled, { css } from "styled-components";
import {
  enterFromLeft,
  enterFromRight,
  exitToLeft,
  exitToRight,
} from "../../Nav.styled";

export const StyledListItem = styled("li")``;

export const StyledNavLink = styled.a`
  width: 100%;
  display: flex;
  cursor: pointer;
  align-items: center;
  justify-content: space-between;
  padding: 16px 28px;
`;

export const StyledSubNav = styled.div``;

export const StyledSubButtonStyles = css`
  width: 100%;
  display: flex;
  cursor: pointer;
  align-items: center;
  padding: 16px 28px;
`;

export const StyledSubTriggerButton = styled.button`
  ${() => StyledSubButtonStyles}
  justify-content: space-between;
`;

export const StyledSubBackButton = styled.button`
  ${() => StyledSubButtonStyles}
  justify-content: flex-start;
  column-gap: 10px;
  border-bottom: 1px solid ${() => theme.colors.grey5};
`;

export const StyledSubContent = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  width: 100%;
  overflow: scroll;
  animation-duration: 250ms;
  animation-timing-function: ease;
  transform-origin: top center;
  width: 100%;
  background-color: white;
  overflow: hidden;
  @media only screen and (min-width: 600px) {
    width: auto;
  }
  &[data-motion="from-start"] {
    animation-name: ${enterFromLeft};
  }
  &[data-motion="from-end"] {
    animation-name: ${enterFromRight};
  }
  &[data-motion="to-start"] {
    animation-name: ${exitToLeft};
  }
  &[data-motion="to-end"] {
    animation-name: ${exitToRight};
  }
`;
