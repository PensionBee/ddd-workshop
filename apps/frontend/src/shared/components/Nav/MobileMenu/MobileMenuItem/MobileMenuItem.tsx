import type { NavMenuItem } from "../../Nav.types";
import { StyledListItem, StyledNavLink } from "./MobileMenuItem.styled";

export const MobileMenuItem: React.FC<NavMenuItem> = ({
  href,
  children,
  ...props
}) => {
  return (
    <StyledListItem data-rc="MobileMenuItem" {...props}>
      <StyledNavLink href={href}>{children}</StyledNavLink>
    </StyledListItem>
  );
};
