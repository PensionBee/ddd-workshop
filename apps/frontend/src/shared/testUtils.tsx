import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
