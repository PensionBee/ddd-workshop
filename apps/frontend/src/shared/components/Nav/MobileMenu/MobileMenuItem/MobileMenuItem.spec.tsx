import { routerRender } from "@shared/testUtils";
import { MobileMenuItem } from "./MobileMenuItem";

const ITEM_CONTENT = "Item Content";
const EXAMPLE_LINK = "/example";

describe("MobileMenuItem", () => {
  it("should render correctly", () => {
    const { getByText, asFragment } = routerRender(
      <MobileMenuItem>{ITEM_CONTENT}</MobileMenuItem>
    );

    expect(getByText(ITEM_CONTENT)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with href", () => {
    const { getByText, asFragment } = routerRender(
      <MobileMenuItem href={EXAMPLE_LINK}>{ITEM_CONTENT}</MobileMenuItem>
    );

    expect(getByText(ITEM_CONTENT).closest("a")).toHaveAttribute(
      "href",
      EXAMPLE_LINK
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
