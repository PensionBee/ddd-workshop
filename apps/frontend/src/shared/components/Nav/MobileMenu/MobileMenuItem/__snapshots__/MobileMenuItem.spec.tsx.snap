// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileMenuItem should render correctly 1`] = `
<DocumentFragment>
  <li
    class="sc-bcXHqe"
    data-rc="MobileMenuItem"
  >
    <a
      class="sc-gswNZR bshdub"
    >
      Item Content
    </a>
  </li>
</DocumentFragment>
`;

exports[`MobileMenuItem should render with href 1`] = `
<DocumentFragment>
  <li
    class="sc-bcXHqe"
    data-rc="MobileMenuItem"
  >
    <a
      class="sc-gswNZR bshdub"
      href="/example"
    >
      Item Content
    </a>
  </li>
</DocumentFragment>
`;
